Scaladocs for new jdk converters, move Accumulators to jdk package

lrytz opened the pull request.

lrytz committed:
Improve scaladocs for steppers

lrytz committed:
StreamConverters scaladoc

lrytz commented:
I put it on RC1, as it will be no-risk to merge once it's ready. But it's not a blcoker.

NthPortal commented:
@lrytz any reason not to label this as a doc PR? (I'll add it and you can remove it if there's a reason why not)

lrytz commented:
Thanks, I keep forgetting to add labels.

lrytz committed:
Make Stepper.trySplit/estimateSize/characteristics public

Having them `private[collection]` is not a good idea because users
cannot implement their own steppers outside the `collection` package
while keeping the same signature.

They are accessed from the outside, for example DoubleStepperSpliterator
accesses `s.characteristics`, so `protected` doesn't work.

lrytz committed:
Move Accumulators to scala.jdk

lrytz committed:
Scaladocs for Accumulator

lrytz commented:
This is ready for review, @adriaanm @szeiger. I included the move of `Accumulator` to `scala.jdk` here.

szeiger commented on src/library/scala/jdk/Accumulator.scala:
```diff
@@ -10,25 +10,72 @@
- * Base class to share code between the [[AnyAccumulator]] class (for reference types) and the manual
- * specializations [[IntAccumulator]], [[LongAccumulator]] and [[DoubleAccumulator]].
- */
+/** Accumulators is a mutable sequences with two distinct features:
```
"Accumulators are mutable sequences"?

szeiger commented on src/library/scala/jdk/Accumulator.scala:
```diff
@@ -10,25 +10,72 @@
+  *
+  * Accumulators can hold more than `Int.MaxValue` elements. They have a [[sizeLong]] method that
+  * returns the size as a `Long`. Note that certain operations defined in [[scala.collection.Seq]]
+  * are implemented using [[length]], so they will not work correctly if for large accumulators.
```
extraneous "if"

szeiger commented on src/library/scala/jdk/Accumulator.scala:
```diff
@@ -10,25 +10,72 @@
+  * returns the size as a `Long`. Note that certain operations defined in [[scala.collection.Seq]]
+  * are implemented using [[length]], so they will not work correctly if for large accumulators.
+  *
+  * The [[Accumulator]] class is a base class to share code between [[AnyAccumulator]] (for
```
It's a bit strange that the type intended for `AnyRef` is called `AnyAccumulator`

lrytz replied:
Yeah, I went back and forth a few times here ;-)

First, there's a correspondence with Steppers, we also have `AnyStepper`, `IntStepper`.

Then, I'd like to keep the supertypes to be `Stepper` and `Accumulator`, so we can say "`IterableOnce.stepper` returns a `Stepper`" (and not "a `StepperBase`"), and so that we can write `someCollection.to(Accumulator)`, where `Accumulator` is the companion of the supertype, which is converted to the factory of the actual implementation according to the element type.

We could say `AnyRefAccumulator`, but it can also hold boxed primitives, there's no bound on the element type. So I'd say it's OK as a compromise.
